Oneandone compute api

alibazlamit opened the pull request:
@nacx Opening this PR for the compute API it is still in development, but just to keep track of progress.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -0,0 +1,214 @@
+         logger.error(ex, ">> failed to provision server. rollbacking..");
+         throw Throwables.propagate(ex);
+      }
+      waitServerUntilAvailable.apply(server);
```
There is no need to wait here. Jclouds will already wait for the server to be running, so let's better remove this.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -0,0 +1,214 @@
+                 .hardware(hardwareRequest)
+                 .applianceId(image.getId())
+                 .dataCenterId(dataCenterId)
+                 .password(password)
```
Is it mandatory to force a password login? What about ssh key access?

alibazlamit replied:
It is not mandatory, do we want to depend completely on ssh and remove the password option?

nacx replied:
We need to configure SSH or Password depending on which credentials the user configured in the template options.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -0,0 +1,214 @@
+
+   @Override
+   public void destroyNode(String id) {
+      waitServerUntilAvailable.apply(getNode(id));
```
Is this really needed here and in the next state-change methods?

alibazlamit replied:
Yes it is needed the API wont respond if the server is not available.

nacx replied:
Is "available" == powered on"? Because the predicate checks for that status.

alibazlamit replied:
I have made another one, now we have wait server until running and wait server until ready,
ready means its either turned on/off and not having any operation in progress, running means the server is powered on.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/config/OneAndOneComputeServiceContextModule.java:
```diff
@@ -0,0 +1,143 @@
+      bind(new TypeLiteral<ComputeServiceAdapter<Server, HardwareFlavour, org.apache.jclouds.oneandone.rest.domain.Image, Location>>() {
+      }).to(OneandoneComputeServiceAdapter.class);
+
+      bind(TemplateBuilderImpl.class).to(ArbitraryCpuRamTemplateBuilderImpl.class);
```
Can users set custom CPU/RAM values even if there are no such predefined hardware flavours in OneAndOne? If not, remove this binding.

alibazlamit replied:
Yes users can set custom CPU/RAM values.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -0,0 +1,214 @@
+
+   @Override
+   public void resumeNode(String id) {
+      waitServerUntilAvailable.apply(getNode(id));
```
This won't work? You wait until the node is powered on when you want to resume it. This will take forever.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/function/ImageToImage.java:
```diff
@@ -0,0 +1,97 @@
+   }
+
+   private boolean is64Bit(int architecture) {
+      switch (architecture) {
```
Just `return architecture != 32;`

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/function/HardwareFlavourToHardware.java:
```diff
@@ -0,0 +1,64 @@
+
+   @Override
+   public Hardware apply(HardwareFlavour from) {
+      int MinRamSize = (int) from.hardware().ram();
```
Start variable name in lowercase

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/function/ServerToNodeMetadata.java:
```diff
@@ -0,0 +1,173 @@
+            size += storage.size();
+            volumes.add(fnVolume.apply(storage));
+         }
+      }
```
Always building a custom hardware object? This seems not right. You should take into account prefedined flavours.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/function/ServerToNodeMetadata.java:
```diff
@@ -0,0 +1,173 @@
+              .build();
+
+      // Collect ips
+      List<String> addresses = fnCollectIps.apply(server.ips());
```
No need for that dummy function in the constructor. Remove it and change this to something like:

```java
List<String> addresses = Lists.transform(new Function<ServerIp, String>() {
   @Override public String apply(ServerIp in) {
      return in.ip();
   }
});
```

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/function/ServerToNodeMetadata.java:
```diff
@@ -0,0 +1,173 @@
+      }
+   }
+
+   static OperatingSystem mapOsType(OSFamliyType osType) {
```
This is a duplicate of the one in the Image transformation function. Reuse it.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/config/OneAndOneHttpApiModule.java:
```diff
@@ -57,15 +57,14 @@ protected void configure() {
    @Provides
-   @Named(POLL_PREDICATE_SERVER)
-   Predicate<Server> provideServerReadyPredicate(final OneAndOneApi api, OneAndOneConstants constants) {
+   Predicate<Server> provideServerReadyPredicate(final OneAndOneApi api, ComputeConstants constants) {
```
Do we have now two different predicates to check if a server is ready?

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/Location.java:
```diff
@@ -0,0 +1,61 @@
+   ES("es", "Spain"),
+   GB("gb", "UNITED KINGDOM"),
+   UNRECOGNIZED("unrecognized", "Unrecognized location"),
+   MOCK("mock", "Mock");
```
What are the unrecognized and mock locations for? Can we remove them?

alibazlamit replied:
I use mock for the mock tests,  unrecognized could be removed.

nacx replied:
Let's use mock tests with real values but avoid putting test code in the production code.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/Provisionable.java:
```diff
@@ -0,0 +1,36 @@
+ * Marker interface for {@link org.jclouds.profitbricks.domain.Image} and
+ * {@link org.jclouds.profitbricks.domain.Snapshot}
+ */
+public interface Provisionable {
```
Remove this class.

nacx commented on oneandone/src/test/java/org/apache/jclouds/oneandone/rest/compute/OneAndOneComputeServiceLiveTest.java:
```diff
@@ -0,0 +1,114 @@
+      ImmutableSet.Builder<Module> modules = ImmutableSet.builder();
+      modules.addAll(super.setupModules());
+      return modules.build();
+   }
```
Remove this method.

nacx commented on oneandone/src/test/java/org/apache/jclouds/oneandone/rest/compute/OneAndOneComputeServiceLiveTest.java:
```diff
@@ -0,0 +1,114 @@
+   @Override
+   public void testOptionToNotBlock() throws Exception {
+      // OneAndOne implementation intentionally blocks until the node is 'AVAILABLE'
+   }
```
There should be no need for that, so this method could be removed.

nacx commented on oneandone/src/test/java/org/apache/jclouds/oneandone/rest/compute/OneAndOneComputeServiceLiveTest.java:
```diff
@@ -0,0 +1,114 @@
+   @Override
+   protected void checkOsMatchesTemplate(NodeMetadata node) {
+      // Not enough description from API to match template
+   }
```
Can you provide an example response for the getImage call? Isn't anything there we can use to determine the OS?

nacx commented on oneandone/src/test/java/org/apache/jclouds/oneandone/rest/compute/OneAndOneTemplateBuilderLiveTest.java:
```diff
@@ -0,0 +1,43 @@
+      ImmutableSet.Builder<Module> modules = ImmutableSet.builder();
+      modules.addAll(super.setupModules());
+      return modules.build();
+   }
```
Remove this method.

nacx commented on oneandone/src/test/java/org/apache/jclouds/oneandone/rest/compute/baseclass.java:
```diff
@@ -0,0 +1,1001 @@
+   protected Module getSshModule() {
+      throw new IllegalStateException("ssh is required for this test!");
+   }
+}
```
Remove this file.

nacx commented on oneandone/src/test/java/org/apache/jclouds/oneandone/rest/compute/OneAndOneComputeServiceLiveTest.java:
```diff
@@ -0,0 +1,114 @@
+
+@Test(groups = "live", singleThreaded = true, testName = "OneAndOneComputeServiceLiveTest")
+public class OneAndOneComputeServiceLiveTest extends BaseComputeServiceLiveTest {
+//    public class OneAndOneComputeServiceLiveTest extends baseclass {
```
Remove this comment

nacx commented on oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/DataCenterApiLiveTest.java:
```diff
@@ -0,0 +1,68 @@
+      Assert.assertTrue(dataCenters.size() > 0);
+   }
+
+   @Test(dependsOnMethods = "testList")
```
No need for this dependency

alibazlamit commented:
@nacx updated with the reviewed items, still having the same issue with live tests, any idea?

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -0,0 +1,217 @@
+
+      }
+
+      return new NodeAndInitialCredentials<Server>(server, server.id(), serverCredentials);
```
In this case there is no need to set the credentials manually here. jclouds will populate the `defaultCredentials` for the image, or the ones provided in the template options automatically. Just pass `null`.

Regarding the default credentials for the images, do they come with some default user and a default password? Or are cloud-init images or similar that create users and passwords on-the-fly?

alibazlamit replied:
If we don't pass an rsakey nor a password  a password will be generated and returned within the responses from the API

nacx replied:
In that case, would it make sense to set it to null if the user configured the credentials explicitly, otherwise, read that response and manually set the generated credentials here?

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -0,0 +1,217 @@
+
+   @Override
+   public void suspendNode(String id) {
+      waitServerUntilRunning.apply(getNode(id));
```
What if we call this on an already suspended node? Will it block until it times out?
Perhaps it is a better idea to validate the state as a precondition and fail, or not validate it at all and let the it fail in the server side.

alibazlamit replied:
ill just validate that the server is available.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/config/OneAndOneComputeServiceContextModule.java:
```diff
@@ -0,0 +1,179 @@
+            return false;
+         } else {
+            return true;
+         }
```
Try to avoid conditionals like `if (condition) return true else return false`. Just `return condition`.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/config/OneAndOneComputeServiceContextModule.java:
```diff
@@ -0,0 +1,179 @@
+   }
+
+   @Singleton
+   public static class ComputeConstants {
```
Get rid of this class and use the existing [PollPeriod](https://github.com/jclouds/jclouds/blob/master/compute/src/main/java/org/jclouds/compute/reference/ComputeServiceConstants.java#L73-L82) and [Timeouts](https://github.com/jclouds/jclouds/blob/master/compute/src/main/java/org/jclouds/compute/reference/ComputeServiceConstants.java#L94-L105).

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/config/OneAndOneComputeServiceContextModule.java:
```diff
@@ -0,0 +1,179 @@
+
+      bind(new TypeLiteral<Function<HardwareFlavour, Hardware>>() {
+      }).to(HardwareFlavourToHardware.class);
+   }
```
If the public key configured by the user can be injected using the API (as it seems), then add [this binding](https://github.com/jclouds/jclouds/blob/master/providers/google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/config/GoogleComputeEngineServiceContextModule.java#L121), so jclouds does not try to upload it as part of a bootstrap ssh script.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/function/ServerToNodeMetadata.java:
```diff
@@ -0,0 +1,155 @@
+            }
+         }
+         // Build hardware
+         String id = String.format("cpu=%f,ram=%d,disk=%f", server.hardware().coresPerProcessor(), (int) server.hardware().ram(), size);
```
Generate the ID with [this](https://github.com/jclouds/jclouds/blob/master/compute/src/main/java/org/jclouds/compute/util/AutomaticHardwareIdSpec.java#L60).

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/config/OneAndOneProperties.java:
```diff
@@ -19,6 +19,7 @@
 public final class OneAndOneProperties {
 
    public static final String POLL_PREDICATE_SERVER = "jclouds.oneandone.rest.predicate.server";
+   public static final String POLL_PREDICATE_SERVER_RUNNING = "jclouds.oneandone.rest.predicate.serveron   ";
```
Remove trailing spaces

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/Location.java:
```diff
@@ -0,0 +1,60 @@
+            return location;
+         }
+      }
+      return US;
```
Better throw an `IllegalArgumentException` instead.

alibazlamit commented:
Changed applied.

Here is the full stack trace of the problem:
```Multiple entries with same value: es=https://cloudpanel-api.1and1.com/v1/ and us=https://cloudpanel-api.1and1.com/v1/
java.lang.IllegalArgumentException
	at com.google.common.collect.ImmutableMap.checkNoConflict(ImmutableMap.java:150)
	at com.google.common.collect.RegularImmutableBiMap.<init>(RegularImmutableBiMap.java:131)
	at com.google.common.collect.ImmutableBiMap.copyOf(ImmutableBiMap.java:210)
	at org.jclouds.suppliers.SupplyKeyMatchingValueOrNull.get(SupplyKeyMatchingValueOrNull.java:53)
	at org.jclouds.rest.suppliers.MemoizedRetryOnTimeOutButNotOnAuthorizationExceptionSupplier$SetAndThrowAuthorizationExceptionSupplierBackedLoader.load(MemoizedRetryOnTimeOutButNotOnAuthorizationExceptionSupplier.java:75)
	at org.jclouds.rest.suppliers.MemoizedRetryOnTimeOutButNotOnAuthorizationExceptionSupplier$SetAndThrowAuthorizationExceptionSupplierBackedLoader.load(MemoizedRetryOnTimeOutButNotOnAuthorizationExceptionSupplier.java:57)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3524)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2317)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2280)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2195)
	at com.google.common.cache.LocalCache.get(LocalCache.java:3934)
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3938)
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4821)
	at org.jclouds.rest.suppliers.MemoizedRetryOnTimeOutButNotOnAuthorizationExceptionSupplier.get(MemoizedRetryOnTimeOutButNotOnAuthorizationExceptionSupplier.java:150)
	at org.jclouds.location.suppliers.implicit.OnlyLocationOrFirstRegionOptionallyMatchingRegionId.get(OnlyLocationOrFirstRegionOptionallyMatchingRegionId.java:59)
	at org.jclouds.location.suppliers.implicit.OnlyLocationOrFirstRegionOptionallyMatchingRegionId.get(OnlyLocationOrFirstRegionOptionallyMatchingRegionId.java:44)
	at org.jclouds.rest.suppliers.MemoizedRetryOnTimeOutButNotOnAuthorizationExceptionSupplier$SetAndThrowAuthorizationExceptionSupplierBackedLoader.load(MemoizedRetryOnTimeOutButNotOnAuthorizationExceptionSupplier.java:75)
	at org.jclouds.rest.suppliers.MemoizedRetryOnTimeOutButNotOnAuthorizationExceptionSupplier$SetAndThrowAuthorizationExceptionSupplierBackedLoader.load(MemoizedRetryOnTimeOutButNotOnAuthorizationExceptionSupplier.java:57)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3524)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2317)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2280)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2195)
	at com.google.common.cache.LocalCache.get(LocalCache.java:3934)
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3938)
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4821)
	at org.jclouds.rest.suppliers.MemoizedRetryOnTimeOutButNotOnAuthorizationExceptionSupplier.get(MemoizedRetryOnTimeOutButNotOnAuthorizationExceptionSupplier.java:150)
	at org.jclouds.compute.domain.internal.TemplateBuilderImpl.build(TemplateBuilderImpl.java:688)
	at org.jclouds.compute.internal.BaseTemplateBuilderLiveTest.testAutoGeneratedHardwareWithOnlyMinCoresMatchHardwareProfile(BaseTemplateBuilderLiveTest.java:282)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:696)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:882)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1189)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:124)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -107,7 +106,8 @@
                  .name(name)
                  .hardware(hardwareRequest)
-                 .rsaKey(options.getPublicKey())
+                 .rsaKey(privateKey)
```
According to the docs, this should be the public key

nacx commented:
@alibazlamit I've had a look at the test failure and it is caused by a jclouds issue. I've opened it as [JCLOUDS-1227](https://issues.apache.org/jira/browse/JCLOUDS-1227). Usually providers configuring the regions via context properties have a different endpoint for each region, so this has not appeared before, and the ProfitBricks provider uses "zones" instead of regions, and the issue does not appear there.

I'll submit a patch in short and I'll ping you once the updated snapshots are published so you can run the tests again.

nacx commented:
See: https://github.com/jclouds/jclouds/pull/1053

nacx commented:
@alibazlamit I've just merged the PR. As soon as the following builds finishes, the updated SNAPSHOTS will be available and you will be able to run the tests again:
https://jclouds.ci.cloudbees.com/job/jclouds/3525/

nacx commented:
@alibazlamit Have you been able to run the tests again, now that the snapshots are in place?

alibazlamit commented:
@nacx Yes i was able to now i am running the live tests and tackling issues along the way.

alibazlamit commented:
@nacx Could you please help me understand better the `RateLimitRetryHandler` and the `RateLimitExceededException` i have implemented both since i was getting too many request from the API, my guess is somehow the failed requests gets a retry after the reset period passes, but running the live tests the testConcurrentUseOfComputeServiceToCreateNodes seems to get stuck and after 20 minutes it  timesout, maybe the implementation of the either the above is missing something?

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/OneAndOneProviderMetadata.java:
```diff
@@ -66,6 +69,7 @@ public static Properties defaultProperties() {
       properties.put(POLL_MAX_PERIOD, 2L * 10L);
       properties.put(PROPERTY_SO_TIMEOUT, 60000 * 5);
       properties.put(PROPERTY_CONNECTION_TIMEOUT, 60000 * 5);
+      properties.put(PROPERTY_MAX_RATE_LIMIT_WAIT, 20);
```
Note that the value is in milliseconds, so under a rate limited environment this is likely to have no effect.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -115,34 +141,113 @@
+              .user(loginUser)
+              .password(password)
+              .privateKey(privateKey)
+              .build();
```
Remove this and don't pass the initial credentials (now it is unused). We can discuss the details of the node access once everything else is in place.

alibazlamit replied:
are these errors related to this issue?

```
no credential or ssh agent found for root on node ED17739E08FCCD5069AF12AD394AAC61
java.lang.IllegalArgumentException
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:148)
	at org.jclouds.compute.functions.CreateSshClientOncePortIsListeningOnNode.apply(CreateSshClientOncePortIsListeningOnNode.java:64)
	at org.jclouds.compute.functions.CreateSshClientOncePortIsListeningOnNode.apply(CreateSshClientOncePortIsListeningOnNode.java:40)
	at org.jclouds.compute.callables.SudoAwareInitManager.init(SudoAwareInitManager.java:67)
	at org.jclouds.compute.callables.RunScriptOnNodeAsInitScriptUsingSsh.init(RunScriptOnNodeAsInitScriptUsingSsh.java:64)
	at org.jclouds.compute.callables.RunScriptOnNodeAsInitScriptUsingSshAndBlockUntilComplete.init(RunScriptOnNodeAsInitScriptUsingSshAndBlockUntilComplete.java:77)
	at org.jclouds.compute.callables.RunScriptOnNodeAsInitScriptUsingSshAndBlockUntilComplete.init(RunScriptOnNodeAsInitScriptUsingSshAndBlockUntilComplete.java:38)
	at org.jclouds.compute.internal.BaseComputeService.runScriptOnNode(BaseComputeService.java:647)
	at org.jclouds.compute.internal.BaseComputeServiceLiveTest.createAndRunAServiceInGroup(BaseComputeServiceLiveTest.java:771)
	at org.jclouds.compute.internal.BaseComputeServiceLiveTest.testCreateAndRunAService(BaseComputeServiceLiveTest.java:741)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:696)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:882)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1189)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:124)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

```

nacx replied:
Then let's keep them for the moment. We'll cleanup the node access configuration in another pull request.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -115,34 +141,113 @@
    @Override
-   public List<HardwareFlavour> listHardwareProfiles() {
-      return api.serverApi().listHardwareFlavours();
+   public List<Hardware> listHardwareProfiles() {
```
Why it is now a hardcoded list instead of calling the API?

alibazlamit replied:
The available flavors provide values less that the live test allows like 0.5 ram value for example, after running the live tests i figured its better to provide flavors this way

nacx replied:
If they are predefined, then they are valid for the provider and we should respectand use them (have you checked that jclouds and the provider use the same units?). In this case I think we'd rather override the assertions in the live tests instead of conditioning the provider logic to them, and get back the old flavor logic.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -115,34 +141,113 @@
+            SingleServerAppliance image = SingleServerAppliance.create(appliance.id(), appliance.name(), availableDatacenters, appliance.osInstallationBase(),
+                    appliance.osFamily(), appliance.os(), appliance.osVersion(), appliance.osArchitecture(), appliance.osImageType(), appliance.minHddSize(),
+                    appliance.type(), appliance.state(), appliance.version(), appliance.categories(), appliance.eulaUrl());
+            if (image != null) {
```
This check is redundant.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -160,8 +265,18 @@ public Server getNode(String id) {
+         }
+      } catch (Exception ex) {
+         logger.debug("no firewall policies found for %s ...", id);
+      }
```
The logic to delete the firewalls should be better moved to a `OneAndOneComputeService` class that implements the `cleanUpIncidentalResourcesOfDeadNodes` method. Thake the [azurecompute-arm implementation](https://github.com/jclouds/jclouds-labs/blob/master/azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/AzureComputeService.java#L103) as an example.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/function/SingleServerApplianceToImage.java:
```diff
@@ -0,0 +1,86 @@
+            return OsFamily.ARCH;
+         }
+      }
+      return OsFamily.UNRECOGNIZED;
```
Better do something like https://github.com/jclouds/jclouds-labs/pull/353 to consider all known OsFamilies.

nacx replied:
Can we change this to use a similar approach?

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/strategy/AssignDataCenterToTemplate.java:
```diff
@@ -0,0 +1,89 @@
+            DataCenter details = api.dataCenterApi().get(input.id());
+            return details != null && template.getLocation().getId().equals((details.countryCode().toLowerCase()));
+         }
+      }, null);
```
Don't use `null` as a default value and let it fail, since it is mandatory.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/handlers/AliRateLimitRetryHandler.java:
```diff
@@ -0,0 +1,128 @@
+ * requests until they are known to succeed.
+ */
+@Beta
+public abstract class AliRateLimitRetryHandler implements HttpRetryHandler {
```
Remove this class

nacx replied:
Remove it :)

nacx commented:
Many conceptual changes in the last commit! I still need to have a deeper look at the hardware profiles and image matching, but here is a first review.

>Could you please help me understand better the RateLimitRetryHandler.

You implemented it right, but it is not retrying as far as I can see, because you have defined a very small max wait period. basically the handler will inspect the headers and determine when the next available request can be sent. If that interval is too much, then it aborts and returns the 429 error and propagates the exception; if not, waits the required time and sends the request again. You need to set the max rate limit wait to a higher value, such as one or two minutes and see.

alibazlamit commented:
@nacx i am having an issue with `testOptionToNotBlock` its throwing back the error below, appearntly the assertion fails when it checks if the server is not running, i assume i have to pass some value to determine wither the server should be turned on after creation or not, how do you think i should proceed with this?

```{id=5808705513F553D7FB7B7DACF9C3F49B, providerId=5808705513F553D7FB7B7DACF9C3F49B, name=oneandone-block-425, location={scope=REGION, id=de, description=de, parent=oneandone, iso3166Codes=[DE-DEU]}, group=oneandone-block, os={family=ubuntu, version=12.04, description=ubuntu1204-64min, is64Bit=true}, status=RUNNING[POWERED_ON], loginPort=22, publicAddresses=[82.165.97.250], hardware={id=automatic:cores=1.0;ram=1;disk=20, providerId=automatic:cores=1.0;ram=1;disk=20, name=automatic:cores=1.0;ram=1;disk=20, location={scope=REGION, id=de, description=de, parent=oneandone, iso3166Codes=[DE-DEU]}, processors=[{cores=1.0, speed=1.0}], ram=1024, volumes=[{id=76ED0EC78992C2F914E9EB3330F21E43, type=LOCAL, size=20.0, bootDevice=true, durable=true}], hypervisor=kvm, supportsImage=ALWAYS_TRUE}, loginUser=root}
java.lang.AssertionError
	at org.jclouds.compute.internal.BaseComputeServiceLiveTest.testOptionToNotBlock(BaseComputeServiceLiveTest.java:864)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:696)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:882)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1189)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:124)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

nacx commented:
> am having an issue with testOptionToNotBlock its throwing back the error below, appearntly the assertion fails when it checks if the server is not running, i assume i have to pass some value to determine wither the server should be turned on after creation or not, how do you think i should proceed with this?

The idea is that the method in the compute service adapter that creates the nodes should trigger the server creation without waiting for the server to be ready. This way users can use the TemplateOptions to configure if they want to block or not. However, looking at the implementation it looks like we need the server to be ready in order to set the firewalls afterwards. Is that true? Can be the firewalls be configured earlier so we can avoid waiting for the server? If not, then let's override this test as in [profitbricks-rest](https://github.com/jclouds/jclouds-labs/blob/master/profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/compute/ProfitBricksComputeServiceLiveTest.java#L71-L74).

alibazlamit commented:
I will have to override it, the server needs to be on.

alibazlamit commented:
@nacx I have updated the code with a few things like the clean up process and returned the hardware flavors, i have one question about the custom hardware settings, how is it supposed to work when the template builder looks through the API flavors to find a matching template?
Maybe something is missing in the implementation,WDYT?

nacx commented:
@alibazlamit Apologies for the late feedback; I haven't had cycles to have a look at this. I'll try to have a look it tomorrow!

alibazlamit commented:
@nacx it is Ok, i was working on fixing all the live tests, all of them now pass successfully.
Since the last review i have added back the hardware flavors and implemented the cleanup process, made a few changes to the way that the hardware is being converted from flavors and back to hardware.

[Here ](https://gist.github.com/alibazlamit/65611afafd7f8b1361fe6a6bad5b6e1d)is the gist of the test results.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/strategy/AssignDataCenterToTemplate.java:
```diff
@@ -0,0 +1,89 @@
+
+      return super.execute(group, count, new TemplateWithDataCenter(template, dataCenter), goodNodes, badNodes,
+              customizationResponses);
+   }
```
If we already use the DataCenter object as the location object in the adapter, we shouldn't need this class, as this list would be performed in the list locations method.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneAndOneComputeService.java:
```diff
@@ -0,0 +1,102 @@
+         }
+      }
+   }
+}
```
The compute service delete node already calls here, so this class shouldn't be necessary.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/strategy/TemplateWithDataCenter.java:
```diff
@@ -0,0 +1,108 @@
+      return delegate.toString();
+   }
+
+}
```
We shouldn't need this if we go for the DataCenter as Location approach.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/config/OneAndOneProperties.java:
```diff
@@ -19,8 +19,9 @@
+   public static final String POLL_PREDICATE_SERVER_RUNNING = "jclouds.oneandone.rest.predicate.serveron";
    public static final String POLL_PREDICATE_PRIVATE_NETWORK = "jclouds.oneandone.rest.predicate.privatenetwork";
-   public static final String AUTH_TOKEN = "oneandone.identity";
+   public static final String AUTH_TOKEN = "oneandone.crednetial";
```
Remove this property and the line that uses it in the live test base class. jclouds already sets it to the right value [here](https://github.com/jclouds/jclouds/blob/master/core/src/test/java/org/jclouds/apis/BaseApiLiveTest.java#L60).

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/Location.java:
```diff
@@ -0,0 +1,60 @@
+            return location;
+         }
+      }
+      throw new IllegalArgumentException();
```
Add a message to the exception.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/ServerAppliance.java:
```diff
@@ -0,0 +1,77 @@
+      "min_hdd_size", "type", "state", "version", "categories", "eula_url"})
+   public static ServerAppliance create(String id, String name, List<String> availableDataCenters, String osInstallationBase, Types.OSFamliyType osFamily, String os,
+           String osVersion, int osArchitecture, Types.OSImageType osImageType, int minHddSize, ApplianceType type, String state, String version, List<String> categories, String eulaUrl) {
+      return new AutoValue_ServerAppliance(id, name, availableDataCenters == null ? ImmutableList.<String>of() : availableDataCenters, osInstallationBase, osFamily, os, osVersion, osArchitecture, osImageType, minHddSize, type, state, version, categories == null ? ImmutableList.<String>of() : categories, eulaUrl);
```
Use `ImmutableList.copyOf` when lists are not null.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/SingleServerAppliance.java:
```diff
@@ -0,0 +1,88 @@
+      "min_hdd_size", "type", "state", "version", "categories", "eula_url"})
+   public static SingleServerAppliance create(String id, String name, List<AvailableDataCenters> availableDataCenters, String osInstallationBase, Types.OSFamliyType osFamily, String os,
+           String osVersion, int osArchitecture, Types.OSImageType osImageType, int minHddSize, Types.ApplianceType type, String state, String version, List<String> categories, String eulaUrl) {
+      return new AutoValue_SingleServerAppliance(id, name, availableDataCenters == null ? ImmutableList.<AvailableDataCenters>of() : availableDataCenters, osInstallationBase, osFamily, os, osVersion, osArchitecture, osImageType, minHddSize, type, state, version, categories == null ? ImmutableList.<String>of() : categories, eulaUrl);
```
Use `ImmutableList.copyOf` when lists are not null.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/util/ServerApplianceAdapter.java:
```diff
@@ -0,0 +1,111 @@
+
+   public ServerApplianceAdapter(com.google.gson.reflect.TypeToken adapterclass) {
+      this.adapterclass = adapterclass;
+      gson = new Gson();
```
Can we get the jclouds `Json` instance injected here instead?

nacx commented on oneandone/src/test/java/org/apache/jclouds/oneandone/rest/compute/OneAndOneComputeServiceLiveTest.java:
```diff
@@ -0,0 +1,116 @@
+
+   @Override
+   protected void checkOsMatchesTemplate(NodeMetadata node) {
+      // Not enough description from API to match template
```
If we change the node transformation function to use the hardware and image caches, we should be able to have this check work. We should definitely not override this.

nacx commented on oneandone/src/test/java/org/apache/jclouds/oneandone/rest/compute/function/CustomHardwareToHardwareTest.java:
```diff
@@ -0,0 +1,82 @@
+import org.testng.annotations.Test;
+
+@Test(groups = "unit", testName = "CustomHardwareToHardwareTest", singleThreaded = true)
+public class CustomHardwareToHardwareTest extends BaseOneAndOneApiMockTest {
```
Do we really need to extend this class? That one should be for mock api tests.

nacx commented on oneandone/src/test/java/org/apache/jclouds/oneandone/rest/compute/function/HardwareFlavourToHardwareTest.java:
```diff
@@ -0,0 +1,84 @@
+import org.testng.annotations.Test;
+
+@Test(groups = "unit", testName = "HardwareFlavourToHardwareTest", singleThreaded = true)
+public class HardwareFlavourToHardwareTest extends BaseOneAndOneApiMockTest {
```
Same comment about inheritance.

nacx commented on oneandone/src/test/java/org/apache/jclouds/oneandone/rest/compute/function/HddToVolumeTest.java:
```diff
@@ -0,0 +1,57 @@
+import org.testng.annotations.Test;
+
+@Test(groups = "unit", testName = "HddToVolumeTest", singleThreaded = true)
+public class HddToVolumeTest extends BaseOneAndOneApiMockTest {
```
Same inheritance comment.

nacx commented on oneandone/src/test/java/org/apache/jclouds/oneandone/rest/compute/function/HardwareFlavourToHardwareTest.java:
```diff
@@ -0,0 +1,84 @@
+              new MockResponse().setBody(stringFromResource("/compute/hardware.json"))
+      );
+
+      org.apache.jclouds.oneandone.rest.domain.HardwareFlavour hardware = api.serverApi().getHardwareFlavour("some-id");
```
There is no need to mock api calls to implement this test.

nacx commented on oneandone/src/test/java/org/apache/jclouds/oneandone/rest/compute/function/HddToVolumeTest.java:
```diff
@@ -0,0 +1,57 @@
+              new MockResponse().setBody(stringFromResource("/compute/volume.json"))
+      );
+
+      org.apache.jclouds.oneandone.rest.domain.Hdd hdd = api.serverApi().getHdd("some-id", "hdd-id");
```
There is no need to mock api calls to implement this test.

nacx commented on oneandone/src/test/java/org/apache/jclouds/oneandone/rest/compute/function/ServerToNodeMetadataTest.java:
```diff
@@ -0,0 +1,184 @@
+import org.testng.annotations.Test;
+
+@Test(groups = "unit", testName = "ServerToNodeMetadataTest", singleThreaded = true)
+public class ServerToNodeMetadataTest extends BaseOneAndOneApiMockTest {
```
Same about inheritance.

nacx commented on oneandone/src/test/java/org/apache/jclouds/oneandone/rest/compute/function/ServerToNodeMetadataTest.java:
```diff
@@ -0,0 +1,184 @@
+      server.enqueue(
+              new MockResponse().setBody(stringFromResource("/compute/server.json"))
+      );
+      Server serverObject = api.serverApi().get("mock");
```
There is no need to mock api calls to implement this test.

nacx commented on oneandone/src/test/java/org/apache/jclouds/oneandone/rest/compute/function/SingleServerApplianceToImageTest.java:
```diff
@@ -0,0 +1,65 @@
+import org.testng.annotations.Test;
+
+@Test(groups = "unit", testName = "ServerApplianceToImageTest", singleThreaded = true)
+public class SingleServerApplianceToImageTest extends BaseOneAndOneApiMockTest {
```
Same inheritance comment.

nacx commented on oneandone/src/test/java/org/apache/jclouds/oneandone/rest/compute/function/SingleServerApplianceToImageTest.java:
```diff
@@ -0,0 +1,65 @@
+              new MockResponse().setBody(stringFromResource("/compute/image.json"))
+      );
+
+      SingleServerAppliance image = api.serverApplianceApi().get("some-id");
```
There is no need to mock api calls to implement this test.

nacx commented on oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/DataCenterApiLiveTest.java:
```diff
@@ -0,0 +1,68 @@
+   public void testList() {
+      dataCenters = dataCenterApi().list();
+      currentDataCenter = dataCenters.get(0);
+      assertNotNull(dataCenters);
```
Check this before calling get.

nacx commented on oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/ServerApplianceApiLiveTest.java:
```diff
@@ -0,0 +1,61 @@
+   public void testList() {
+      appliances = serverApplianceApi().list();
+      currentAppliance = appliances.get(0);
+      Assert.assertTrue(appliances.size() > 0);
```
Assert this before calling get.

alibazlamit commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -0,0 +1,288 @@
+
+         server = api.serverApi().create(serverRequest);
+
+         waitServerUntilAvailable.apply(server);
```
It is required otherwise the create firewall rules would fail.

alibazlamit commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -0,0 +1,288 @@
+   @Override
+   public Iterable<SingleServerAppliance> listImages() {
+      GenericQueryOptions options = new GenericQueryOptions();
+      options.options(0, 0, null, Types.OSImageType.Minimal.toString(), null);
```
Here is an issue with 1&1 there are over 90 server appliance, i could not figure out how to lower down the number or what is the difference, minimal i assumed is a basic version of the image without additional software.

nacx replied:
Is is a problem to list them all? If it is, we could add a context property for a string to configure this search, so users can customize, if they want to filter images, the type of images they're interested in.

alibazlamit commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/OneAndOneProviderMetadata.java:
```diff
@@ -49,12 +53,23 @@ public OneAndOneProviderMetadata(Builder builder) {
+      properties.setProperty(PROPERTY_REGION + ".de." + ISO3166_CODES, "DE-DEU");
+      properties.setProperty(PROPERTY_REGION + ".us." + ISO3166_CODES, "US-USA");
+      properties.setProperty(PROPERTY_REGION + ".es." + ISO3166_CODES, "ES-ESP");
+      properties.setProperty(PROPERTY_REGION + ".gb." + ISO3166_CODES, "GB-GBR");
```
Should i just remove them?

nacx replied:
When the ComptueService lsits the DataCenters to get the locations, yes.

alibazlamit commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/OneandoneComputeServiceAdapter.java:
```diff
@@ -0,0 +1,288 @@
+import org.jclouds.rest.ResourceNotFoundException;
+
+@Singleton
+public class OneandoneComputeServiceAdapter implements ComputeServiceAdapter<Server, HardwareFlavour, SingleServerAppliance, Location> {
```
Can you help me figure out how to make the location supplier load datacetners as locations its loading regions from properties as locations.

nacx replied:
You need to add something [like this](https://github.com/jclouds/jclouds/blob/master/providers/digitalocean2/src/main/java/org/jclouds/digitalocean2/compute/config/DigitalOcean2ComputeServiceContextModule.java#L89-L90) to your compute service module, and you'll have to create a function that transforms a DataCenter into a Location (and obviously implement in the adapter the list locations method).

You can keep this binding. It just configures which location to use when the user does not set it. That is the [default implementation](https://github.com/jclouds/jclouds/blob/master/core/src/main/java/org/jclouds/location/suppliers/ImplicitLocationSupplier.java#L25) though, so the binding is redundant and can be removed.

alibazlamit commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/function/ServerToNodeMetadata.java:
```diff
@@ -0,0 +1,216 @@
+                 .hypervisor("kvm")
+                 .volumes(volumes)
+                 .location(location)
+                 .build();
```
I am thinking of making something similar to [this](https://github.com/jclouds/jclouds/blob/4bbca9edf943852ce1ea5aa579fa0554f770a3ea/providers/google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/domain/internal/GoogleComputeEngineArbitraryCpuRamTemplateBuilderImpl.java), WDYT?

alibazlamit replied:
I was thinking about creating something [like ](https://github.com/jclouds/jclouds/blob/4bbca9edf943852ce1ea5aa579fa0554f770a3ea/providers/google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/domain/internal/GoogleComputeEngineArbitraryCpuRamTemplateBuilderImpl.java)this and adding all the custom hardware logic in that class, is that the needed approach?

nacx replied:
I don't think so. That is there because Google has its own way of deploying custom hardware profiles, encoding the values in the "size uri".
Let's leave this as-is in this PR. It's not a critical thing to change.

alibazlamit commented:
@nacx i have done all of the changes above, i noticed one issue with the `RateLimitRetryHandler` it does not seem to work right or it does not fit the way the API works, the API sends in each response header that looks like this 
```
X-Rate-Limit-Limit: 25
X-Rate-Limit-Remaining: 24
```
Is there a way to check for the Remaining flag in case it comes down to 1 to make it wait for a minute, each minute we are allowed 25 requests, after we pass the limit the API sends an error response and in the headers there is the `X-Rate-Limit-Reset` that states how many seconds to wait for a new limit, in this case its too late because the request failed already.

nacx commented:
This is the expected behavior. When you exceed the rate limit, the API returns a 429 error, and then the retry handler is invoked. It is OK that the request has failed. It is the retry handler that will decide if that failed request can be retried in a while
The current implementation properly reads the time to wait, so it would be working fine.

What errors are you experiencing? Aren't you getting a 429 response error when the rate limit is exceeded?

alibazlamit commented:
This is an example of the error i get, the problems is the flow breaks and tests fail, i have this property set `properties.put(PROPERTY_MAX_RATE_LIMIT_WAIT, 60 * 1000);` it is passed to rate limit class, one minute is the maximum time needed to wait for a new limit, i am not sure what is wrong but it seems though as the failed requests are not being retried.

```
HTTP/1.1 429 Too Many Requests
Date: Fri, 17 Feb 2017 17:37:38 GMT
Server: Apache
Error-Msg: TOO_MANY_REQUESTS
X-Rate-Limit-Reset: 21
Strict-Transport-Security: max-age=63072000; includeSubDomains
Access-Control-Allow-Origin: *
Access-Control-Allow-Headers: X-API-TOKEN, X-USER
Access-Control-Allow-Methods: POST
Access-Control-Expose-Headers: Error-Msg
Connection: close
Content-Type: application/json
Set-Cookie: visid_incap_402159=9rMEgI4gTKWn9++6PSKzD+o0p1gAAAAAQUIPAAAAAADaOyYtVzQ8/Mk0MKear4J+; expires=Sat, 17 Feb 2018 15:49:54 GMT; path=/; Domain=.1and1.com
Set-Cookie: incap_ses_260_402159=c9RdblutIGDuZNf9YeqbA+o0p1gAAAAAg/T6oD9pkSAJrmSHgllx2w==; path=/; Domain=.1and1.com
X-Iinfo: 7-38715022-38715034 NNNN CT(38 86 0) RT(1487353056950 49) q(0 0 2 -1) r(100 100) U5
X-CDN: Incapsula
Content-Length: 115

{"type":"TOO_MANY_REQUESTS","message":"The maximum number of requests per minute has been exceeded.","errors":null}
```

nacx commented:
That's the problem. The `PROPERTY_MAX_RATE_LIMIT_WAIT` controls how much time you are willing to wait for a rate limited request to be retried. In OneAndOne, it seems that in most cases you'll have to wait, at least, one minute, so the value you are setting does not seem OK. You should set it to 5 minutes (300000) or so.

You know that upon limit consumption, you'll have to delay requests, at least, for a minute. Let's configure the maximum time you are willing to wait for them to a higher value.

Does it make sense now?

alibazlamit commented:
Changing the property works, gist [here ](https://gist.github.com/alibazlamit/65611afafd7f8b1361fe6a6bad5b6e1d) and all changes requested are added.

Thanks @nacx !

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/function/ServerToNodeMetadata.java:
```diff
@@ -28,52 +28,50 @@
 public class ServerToNodeMetadata implements Function<Server, NodeMetadata> {
 
    private final Supplier<Set<? extends Location>> locations;
+   private final Supplier<Map<String, ? extends Hardware>> hardwareFalvors;
```
typo: `Falvors`

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/domain/SingleServerAppliance.java:
```diff
@@ -70,7 +70,63 @@
+
+      public abstract Builder eulaUrl(String eulaUrl);
+
+      public abstract SingleServerAppliance build();
```
Use [this pattern](https://github.com/jclouds/jclouds-labs/blob/master/azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/domain/AvailabilitySet.java#L82-L90) to enforce immutable collections in builders.

nacx commented on oneandone/src/main/java/org/apache/jclouds/oneandone/rest/handlers/OneAndOneRateLimitRetryHandler.java:
```diff
@@ -41,6 +41,6 @@
 
    public static long millisUntilNextAvailableRequest(long secondsToNextAvailableRequest) {
-      return secondsToNextAvailableRequest * 1000;
+      return secondsToNextAvailableRequest * 8000;
```
This should return the seconds until the next available request. The previous value looked OK, as it configures the exact value returned by the provider in the response headers.

alibazlamit commented:
@nacx Thanks for the reply, i have change those minor issues, pushed and ready.

nacx commented:
Great job! I assume the live tests still pass after having changed the rate limit module.
Could you squash the commits and rebase to the latest master so we can merge this? I'll merge it next week, as we're now in the middle of the release process for 2.0.1.

alibazlamit commented:
Squashed and updated from upstream, after merging i assume this goes to the jclouds live repository?

nacx commented:
What do you mean by "the jclouds live repository"?
BTW, there is one checkstyle failure:
https://jclouds.ci.cloudbees.com/job/jclouds-labs-pull-requests/1691/org.apache.jclouds.labs$oneandone/violations/file/src/test/java/org/apache/jclouds/oneandone/rest/compute/basebase.java/

alibazlamit commented:
Fixed the checkstyle issue.
I meant to be moved from labs to become live and available for public use.

nacx commented:
>I meant to be moved from labs to become live and available for public use.

The providers in *labs* are released in every release, so they are already available to anyone that want to use them. Being in labs, however, means that it is not yet production ready, that there is very active development going on there, and that we don't guarantee backwards compatibility between releases. Just that.

I'd like to see the live tests pass consistently for some time before moving this to the main repo, but it is definitely a candidate to be promoted in the next major. Something that helps speed up this is to have some testing account we can use to test. Some providers donate accounts, for others we use our personal ones. If you have a contact we could reach to talk about this kind of collaboration, please, send me an email!

nacx commented:
Pushed to [master](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/eeaa3c66) and [2.0.x](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/eca63496). Thanks @alibazlamit!
